app/scripts/controllers/permissions/specifications.js

```diff
@@ -130,31 +120,49 @@ export const getPermissionSpecifications = ({
 
       methodImplementation: async (_args) => {
         const accounts = await getAllAccounts();
-        const identities = getIdentities();
+        const internalAccounts = getInternalAccounts();
 
         return accounts.sort((firstAddress, secondAddress) => {
-          if (!identities[firstAddress]) {
-            captureKeyringTypesWithMissingIdentities(identities, accounts);
+          const firstAccount = internalAccounts.find(
+            (internalAccount) =>
+              internalAccount.address.toLowerCase() ===
+              firstAddress.toLowerCase(),
+          );
+
+          const secondAccount = internalAccounts.find(
+            (internalAccount) =>
+              internalAccount.address.toLowerCase() ===
+              secondAddress.toLowerCase(),
+          );
+
+          if (!firstAccount) {
+            captureKeyringTypesWithMissingIdentities(
+              internalAccounts,
+              accounts,
+            );
             throw new Error(`Missing identity for address: "${firstAddress}".`);
-          } else if (!identities[secondAddress]) {
-            captureKeyringTypesWithMissingIdentities(identities, accounts);
+          } else if (!secondAccount) {
+            captureKeyringTypesWithMissingIdentities(
+              internalAccounts,
+              accounts,
+            );
             throw new Error(
               `Missing identity for address: "${secondAddress}".`,
             );
           } else if (
-            identities[firstAddress].lastSelected ===
-            identities[secondAddress].lastSelected
+            firstAccount.metadata?.lastSelected ===
+            secondAccount.metadata?.lastSelected
```

Comment:
Is the `?.` necessary here and in the next line, since you don't use it below?

``` suggestion
            firstAccount.metadata.lastSelected ===
            secondAccount.metadata.lastSelected
```

Reply:
No, is not. Solved in [dfae965](https://github.com/MetaMask/metamask-extension/pull/21437/commits/dfae965fbde36477fc9e83829a64760b99b6725a)

```ts
const InternalAccountStruct = object({
  ...KeyringAccountStruct.schema,
  metadata: object({
    snap: optional(...),
    name: string(),
    keyring: object(...),
  }),
});
```